### PR TITLE
Add edit profile flow

### PR DESCRIPTION
**Pull request by ahmarzoukk**
This PR handles issue https://github.com/yebrahim/codersquare/issues/30. It mainly does the following: 

- Adds a backend endpoint for updating profiles:  `PATCH api/v1/users`.  

It requires the following in the request body:
```json
{
 "userName": "string",
 "firstName": "string",
 "lastName": "string",
}
```
Users can update any of the above-mentioned attributes and only theirs as the updates reference the already logged-in user. 

- Updates the UI slightly to allow users to edit each of the previously mentioned profile attributes. 

The new UI changes are screenshotted below based on the related use cases:

Users viewing other users' profiles:
Users viewing other users' profiles
Users viewing their own profiles:
Users viewing their own profile
Users editing their own profiles:
Users editing their own profile

- Updates the build script at the root package.json to install the dev dependencies that it uses (e.g. `npm-run-all`)

### CODE REVIEW

File: web/src/components/navbar.tsx [hunk 1]
```diff
@@ -19,11 +20,16 @@ export const Navbar = () => {
     signOut();
```
Comment (yebrahim) [inline web/src/components/navbar.tsx:20]: need to also set current user to undefined here.

File: web/src/App.tsx [hunk 2]
```diff
@@ -11,22 +13,32 @@ import { ViewPost } from './pages/view-post';
 import { ROUTES } from './routes';
 import { isDev } from './util';
 
+type UserContext = {
+  currentUser: GetCurrentUserResponse;
+  setCurrentUserGlobally: Dispatch<SetStateAction<GetCurrentUserResponse>>;
```
Comment (yebrahim) [inline web/src/App.tsx:18]: these should probably be optional?
```
{
  currentUser?: ...
  setCurrentUserGlobally?: ...
}
```
then you could use the type `UserContext` when creating the state below: `useState<UserContext>({});`.

File: web/src/pages/user-profile.tsx [hunk 3]
```diff
@@ -1,16 +1,66 @@
...
+      try {
+        await callEndpoint<UpdateCurrentUserRequest, UpdateCurrentUserResponse>(
+          ENDPOINT_CONFIGS.updateCurrentUser,
+          {
+            userName,
+            firstName,
+            lastName,
+          }
+        );
+        window.location.reload();
```
Comment (yebrahim) [inline web/src/pages/user-profile.tsx:51]: instead of a full reload, we could just call the `setCurrentUserGlobally` callback from the context?

Comment (yebrahim) [reply web/src/pages/user-profile.tsx:51]: also `setUserUpdateError('')` in case of success here.

File: server/src/handlers/userHandler.ts [hunk 4]
```diff
@@ -114,6 +116,26 @@ export class UserHandler {
     });
   };
 
+  public updateCurrentUser: ExpressHandler<UpdateCurrentUserRequest, UpdateCurrentUserResponse> =
```
Comment (yebrahim) [inline server/src/handlers/userHandler.ts:119]: this endpoint allows you to edit other users' username, no?
we need another check that the `existingUser`'s id below is equal to `res.locals.userId`.

**Review by yebrahim — COMMENTED** [review]
thanks for the PR, and sorry for the late review! I appreciate the screenshot as well! :)

**Comment by ahmarzoukk** [comment]
@yebrahim Sorry I got busy lately. Hopefully, I'll be able to address your comments this weeekend.

**Comment by yebrahim** [comment]
@ahmarzoukk no worries, thanks for the contribution. :)

**Comment by yebrahim** [comment]
@ahmarzoukk are you getting back. on this or should I just commandeer it? :)

File: server/src/handlers/userHandler.ts (hunk 4 above)
Comment (ahmarzoukk) [reply server/src/handlers/userHandler.ts:119]: No, it doesn't. All profile updates target the current authenticated use as I do the updates query against `res.locals.userId`.

**Comment by ahmarzoukk** [comment]
@yebrahim This PR had entirely slipped my mind. So sorry for any inconvenience! 

I tried to improve my changes a little bit while I was addressing your comments! Please review!

File: web/src/App.tsx (hunk 2 above)
Comment (ahmarzoukk) [reply web/src/App.tsx]: This didn't work for me as I expect.

File: web/src/components/userContext.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+} from '@codersquare/shared';
+import { useQuery } from '@tanstack/react-query';
+import { ReactNode, createContext, useContext } from 'react';
+
+import { callEndpoint } from '../fetch';
+import { isLoggedIn } from '../fetch/auth';
+
+type UserContext = {
+  currentUser?: GetCurrentUserResponse;
+  refreshCurrentUser?: any;
```
Comment (yebrahim) [inline web/src/components/userContext.tsx:14]: nit: type this as a function, maybe something like `refreshCurrentUser?: () => void`.

**Review by yebrahim — APPROVED** [review]
nice, thanks for getting back on this. lgtm overall, with a few nits but I won't block on them. let me know if you wanna merge as is. :)

**Comment by ahmarzoukk** [comment]
@yebrahim, I addressed your comments. Please feel free to merge once you're happy with the changes.

**Pull request merged**